src/pygama/raw/orca/orca_SIS3316.py

```diff
@@ -0,0 +1,229 @@
+
+import numpy as np
+
+from pygama.raw.orca.orca_header import OrcaHeader
+from pygama.raw.orca.orca_packet import OrcaPacket
+from pygama.raw.raw_buffer import RawBufferLibrary
+
+from .orca_base import OrcaDecoder, get_ccc
+
+log = logging.getLogger(__name__)
+
+
+class ORSIS3316WaveformDecoder(OrcaDecoder):
+    """Decoder for FlashCam ADC data written by ORCA."""
+
+    def __init__(self, header: OrcaHeader = None, **kwargs) -> None:
+        # self.decoder_name = "ORSIS3316WaveformDecoder"
+        # self.orca_class_name = "ORSIS3316Model"
+
+        # store an entry for every event
+        self.decoded_values_template = {
+            "packet_id": {
+                "dtype": "uint32",
+            },
+            "energy": {
+                "dtype": "uint32",
+                "units": "adc",
+            },
+            "energy_first": {
+                "dtype": "uint32",
+            },
+            "timestamp": {
+                "dtype": "uint64",
+                "units": "clock_ticks",
+            },
+            "crate": {
+                "dtype": "uint8",
+            },
+            "card": {
+                "dtype": "uint8",
+            },
+            "channel": {
+                "dtype": "uint8",
+            },
+            "waveform": {
+                "dtype": "uint16",
+                "datatype": "waveform",
+                "wf_len": 65532,  # max value. override this before initalizing buffers to save RAM
+                "dt": 8,  # override if a different clock rate is use
+                "dt_units": "ns",
+                "t0_units": "ns",
+            },
+        }
+
+        # self.event_header_length = 1 #?
+        self.decoded_values = {}
+        self.skipped_channels = {}
+        super().__init__(
+            header=header, **kwargs
+        )  # also initializes the garbage df (whatever that means...)
+
+    def set_header(self, header: OrcaHeader) -> None:
+        self.header = header
+        import copy
+
+        self.decoded_values = copy.deepcopy(self.decoded_values_template)
+        # print("header", self.header['ObjectInfo']['Crates'])
+        # print("cards", self.header['ObjectInfo']['Crates'][0]['Cards'])
+
+        for card_dict in self.header["ObjectInfo"]["Crates"][0]["Cards"]:
+            if card_dict["Class Name"] == "ORSIS3316Model":
+                card = card_dict["Card"]
+                crate = 0
+                for channel in range(0, 16):
+                    ccc = get_ccc(crate, card, channel)
+                    trace_length = card_dict["rawDataBufferLen"]
+                    self.decoded_values[ccc] = copy.deepcopy(
+                        self.decoded_values_template
+                    )
+
+                    if trace_length <= 0 or trace_length > 2**16:
+                        raise Exception(
+                            "SIS3316ORCADecoder Error: invalid trace_length",
+                            trace_length,
+                        )
```

Comment:
```suggestion
                        raise RuntimeError("invalid trace_length: ", trace_length)
```
Please raise this if something unexpected happens during decoding at runtime, `Exception` is too generic. `SIS3316ORCADecoder Error: ` is not needed since it will be clear from the stacktrace. Please fix this also in the rest of the code.